drivers/volume/portworx/portworx.go

```diff
@@ -804,28 +828,104 @@ func (p *portworx) secretRefFromAnnotations(annotations map[string]string) *v1.S
 	return nil
 }
 
-func (p *portworx) getUserVolDriver(annotations map[string]string) (volume.VolumeDriver, error) {
-	if v, ok := annotations[auth_secrets.SecretNameKey]; ok {
-		token, err := auth_secrets.GetToken(&api.TokenSecretContext{
-			SecretName:      v,
-			SecretNamespace: annotations[auth_secrets.SecretNamespaceKey],
-		})
-		if err != nil {
-			return nil, err
+func (p *portworx) mergeAuthLabels(
+	pvc *v1.PersistentVolumeClaim,
+	pv *v1.PersistentVolume,
+	crAnnotations map[string]string,
+) map[string]string {
+	// Precendence order:
+	// PV.Spec.SecretRef labels > PVC annotations > CR annotations
+	authLabels := make(map[string]string)
+	if val, ok := crAnnotations[auth_secrets.SecretNameKey]; ok {
+		authLabels[auth_secrets.SecretNameKey] = val
+	}
+	if val, ok := crAnnotations[auth_secrets.SecretNamespaceKey]; ok {
+		authLabels[auth_secrets.SecretNamespaceKey] = val
+	}
+	if pvc != nil {
+		if val, ok := pvc.Annotations[auth_secrets.SecretNameKey]; ok {
+			authLabels[auth_secrets.SecretNameKey] = val
+		}
+		if val, ok := pvc.Annotations[auth_secrets.SecretNamespaceKey]; ok {
+			authLabels[auth_secrets.SecretNamespaceKey] = val
+		}
+	}
+	nameKey, namespaceKey := p.getPVSecretRefLabels(pvc, pv)
+	if len(nameKey) > 0 && len(namespaceKey) > 0 {
+		authLabels[auth_secrets.SecretNameKey] = nameKey
+		authLabels[auth_secrets.SecretNamespaceKey] = namespaceKey
+	}
+	return authLabels
+}
+
+func (p *portworx) getPVSecretRefLabels(pvc *v1.PersistentVolumeClaim, pv *v1.PersistentVolume) (string, string) {
+	// CSI today supports the following templatized secret references
+	// - csi.storage.k8s.io/provisioner-secret-name: px-user-token
+	// - csi.storage.k8s.io/provisioner-secret-namespace: ${pvc.namespace}
+	// - csi.storage.k8s.io/node-publish-secret-name: px-user-token
+	// - csi.storage.k8s.io/node-publish-secret-namespace: ${pvc.namespace}
+	// - csi.storage.k8s.io/controller-expand-secret-name: px-user-token
+	// - csi.storage.k8s.io/controller-expand-secret-namespace: ${pvc.namespace}
+	// The provisioner secrets would have been the ideal ones to use, but they are not
+	// copied over into the PV object. Hence we are overloading the NodePublish secrets
+	if pvc != nil && pv != nil && pv.Spec.CSI != nil && pv.Spec.CSI.NodePublishSecretRef != nil {
+		// If the secret ref namespace is same as the PVC namespace check if
+		// templatized namespace was provided
+		if pv.Spec.CSI.NodePublishSecretRef.Namespace == pvc.Namespace {
+			sc, err := core.Instance().GetStorageClassForPVC(pvc)
+			if err != nil {
+				logrus.Warnf("Failed to fetch storage class for PVC %v/%v: %v", pvc.Name, pvc.Namespace, err)
+				// Best effort - return what we have
+				return pv.Spec.CSI.NodePublishSecretRef.Name, pv.Spec.CSI.NodePublishSecretRef.Namespace
+			}
+			if val, ok := sc.Parameters["csi.storage.k8s.io/node-publish-secret-namespace"]; ok && val == templatizedNamespace {
```

Comment:
Not sure if we want to add this, but CSI also supports the `pvc.Name` template:
https://kubernetes-csi.github.io/docs/secrets-and-credentials-storage-class.html#createdelete-volume-secret

Users of CSI might expect `{{pvc.Name`}}` as a possibility. If this isn't the case, we can just document accordingly.

Reply:
Sounds good